src/bootstrap/doc.rs

```diff
@@ -522,6 +522,76 @@ impl Step for Std {
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct JsonStd {
+    pub stage: u32,
+    pub target: TargetSelection,
+}
+
+impl Step for JsonStd {
+    type Output = ();
+    const DEFAULT: bool = false;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        let builder = run.builder;
+        run.default_condition(builder.config.json_docs)
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(Std { stage: run.builder.top_stage, target: run.target });
+    }
+
+    /// Build JSON documentation for the standard library crates.
+    ///
+    /// This is largely just a wrapper around `cargo doc`.
+    fn run(self, builder: &Builder<'_>) {
+        let stage = self.stage;
+        let target = self.target;
+        builder.info(&format!("Building JSON documentation for stage{} std ({})", stage, target));
+        let out = builder.json_doc_out(target);
+        t!(fs::create_dir_all(&out));
+        let compiler = builder.compiler(stage, builder.config.build);
+
+        let out_dir = builder.stage_out(compiler, Mode::Std).join(target.triple).join("doc");
+
+        let run_cargo_rustdoc_for = |package: &str| {
+            let mut cargo =
+                builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "rustdoc");
+            compile::std_cargo(builder, target, compiler.stage, &mut cargo);
+
+            cargo
+                .arg("-p")
+                .arg(package)
+                .arg("-Zskip-rustdoc-fingerprint")
+                .arg("--")
+                .arg("-Z")
+                .arg("unstable-options")
+                .arg("--output-format")
+                .arg("json")
+                .arg("--resource-suffix")
+                .arg(&builder.version);
+
+            builder.run(&mut cargo.into());
+        };
+
+        // Only build the following crates. While we could just iterate over the
+        // folder structure, that would also build internal crates that we do
+        // not want to show in documentation. These crates will later be visited
+        // by the rustc step, so internal documentation will show them.
+        //
+        // Note that the order here is important! The crates need to be
+        // processed starting from the leaves, otherwise rustdoc will not
+        // create correct links between crates because rustdoc depends on the
+        // existence of the output directories to know if it should be a local
+        // or remote link.
+        let krates = ["core", "alloc", "std", "proc_macro", "test"];
+        for krate in &krates {
+            run_cargo_rustdoc_for(krate);
+        }
+        builder.cp_r(&out_dir, &out);
```

Comment:
This duplicates a lot of code from `impl Step for Std`. Can you please add a new `fn doc_std(stage: u32, target: TargetSelection, out_dir: &Path, extra_args: &[&str])` function that has everything here except `--output-format json` and use that in both places?